src/bluesky/plans.py

```diff
@@ -18,16 +20,54 @@
 from . import plan_patterns, utils
 from . import plan_stubs as bps
 from . import preprocessors as bpp
-from .utils import Msg, get_hinted_fields
-
-
-def count(detectors, num=1, delay=None, *, per_shot=None, md=None):
+from .protocols import Flyable, Movable, NamedMovable, Readable
+from .utils import (
+    CustomPlanMetadata,
+    Msg,
+    MsgGenerator,
+    ScalarOrIterableFloat,
+    get_hinted_fields,
+)
+
+#: Plan function that can be used for each shot in a detector acquisition involving no actuation
+PerShot = Callable[[Sequence[Readable], Optional[bps.TakeReading]], MsgGenerator]
+
+#: Plan function that can be used for each step in a scan
+PerStep1D = Callable[
+    [Sequence[Readable], Movable, Any, Optional[bps.TakeReading]],
+    MsgGenerator,
+]
+PerStepND = Callable[
+    [
+        Sequence[Readable],
+        Mapping[Movable, Any],
+        Dict[Movable, Any],
+        Optional[bps.TakeReading],
+    ],
+    MsgGenerator,
+]
+PerStep = Union[PerStep1D, PerStepND]
+
+
+def _check_detectors_type_input(detectors):
+    if not isinstance(detectors, Sequence):
+        raise TypeError("The input argument must be either as a list or a tuple of Readable objects.")
+
+
+def count(
+    detectors: Sequence[Readable],
```

Comment:
Slowly resurrecting this change, @tacaswell and @dmgav, how would you feel about making these `set[Readable]` instead of `Sequence[Readable]`? We have use cases that would be easier if it was made clear that these were unique-valued sequences. The counter argument, I suppose, is if they need to be ordered.

Reply:
They do need to be ordered. But maybe we can find another way. Would you describe the use case?

Reply:
We generate json schemas from plans args using pydantic. Using sets leads to `"uniqueitems": true` in the schema, whereas sequences do not. We're investigating feeding the schemas into [json forms](https://jsonforms.io/), which presents very different interfaces for collections with and without `uniqueitems`.

The alternative, I guess, is to have some custom annotation somewhere and post-processing of the JSON schema, which is definitely an option! Or just wrap the plans.

But yes, I can see they need to be ordered, annoying python has no ordered set type...

Reply:
There is internal logic to de-duplicate devices so it is OK if they are not unique.

The thing that is a problem is if there are over-lapping keys in the read between multiple devices and I do not think there is a way to express that in the typing.